File: mobile/lib/domain/usecases/submit_reinforcement_use_case.dart
import '../repositories/i_learner_repository.dart';
import '../repositories/i_lesson_repository.dart';
import '../entities/question.dart';
import '../entities/skill_score.dart';
import '../entities/learning_path_node.dart';
import '../personalization/skill_scoring_engine.dart';
import '../personalization/progress_engine.dart';
import '../personalization/route_engine.dart';

class ReinforcementResult {
  final bool isCorrect;
  final int oldScore;
  final int newScore;
  final int xpAwarded;
  final LearningPath updatedRoadmap;

  const ReinforcementResult({
    required this.isCorrect,
    required this.oldScore,
    required this.newScore,
    required this.xpAwarded,
    required this.updatedRoadmap,
  });
}

class SubmitReinforcementUseCase {
  final ILearnerRepository learnerRepository;
  final ILessonRepository lessonRepository;

  const SubmitReinforcementUseCase(this.learnerRepository, this.lessonRepository);

  Future<ReinforcementResult> execute({
    required String courseId,
    required String skillId,
    required String reinforcementNodeId,
    required Question reinforcementQuestion,
    required int selectedIndex,
  }) async {
    final isCorrect = selectedIndex == reinforcementQuestion.correctIndex;
    final profile = await learnerRepository.getProfile();
    final courseScores = profile.scoresByCourse[courseId] ?? {};
    final currentSkillScore = courseScores[skillId]?.score ?? 41;

    if (!isCorrect) {
      final baseNodes = await lessonRepository.getBaseRoadmapNodes(courseId);
      final currentRoadmap = RouteEngine.adaptRoadmap(
        courseId: courseId,
        baseNodes: baseNodes,
        skillScores: courseScores,
        completedNodeIds: profile.completedNodeIds,
        completedReinforcementIds: profile.completedReinforcementIds,
      );
      return ReinforcementResult(
        isCorrect: false,
        oldScore: currentSkillScore,
        newScore: currentSkillScore,
        xpAwarded: 0,
        updatedRoadmap: currentRoadmap,
      );
    }

    // 1. Calculate new score deterministically
    final newScore = SkillScoringEngine.calculateReinforcementScore(currentSkillScore);

    // 2. Mark reinforcement node as completed
    await learnerRepository.markReinforcementCompleted(reinforcementNodeId);

    // Also mark target node as completed
    final targetNodeId = reinforcementNodeId.replaceAll('reinf_', '');
    await learnerRepository.markNodeCompleted(targetNodeId);

    // 3. Save new score
    final updatedSkillScore = SkillScore(
      skillId: skillId,
      courseId: courseId,
      score: newScore,
      lastUpdated: DateTime.now().millisecondsSinceEpoch,
      masteryLevel: SkillScoringEngine.getMasteryLevel(newScore),
      isWeakestFocus: false,
    );

    await learnerRepository.saveSkillScores(courseId, {
      skillId: updatedSkillScore,
    });

    // 4. Calculate XP idempotently
    final xpCalc = ProgressEngine.calculateNewXP(
      currentXP: profile.xp,
      actionType: 'reinforcement_completion',
      actionId: reinforcementNodeId,
      completedActionIds: profile.completedReinforcementIds,
    );

    if (xpCalc.wasAwarded) {
      final updatedProfile = await learnerRepository.getProfile();
      await learnerRepository.updateProfile(updatedProfile.copyWith(xp: xpCalc.newXP));
    }

    // 5. Adapt updated roadmap
    final finalProfile = await learnerRepository.getProfile();
    final baseNodes = await lessonRepository.getBaseRoadmapNodes(courseId);
    final updatedScores = finalProfile.scoresByCourse[courseId] ?? {};

    final updatedRoadmap = RouteEngine.adaptRoadmap(
      courseId: courseId,
      baseNodes: baseNodes,
      skillScores: updatedScores,
      completedNodeIds: finalProfile.completedNodeIds,
      completedReinforcementIds: finalProfile.completedReinforcementIds,
    );

    return ReinforcementResult(
      isCorrect: true,
      oldScore: currentSkillScore,
      newScore: newScore,
      xpAwarded: xpCalc.wasAwarded ? 30 : 0,
      updatedRoadmap: updatedRoadmap,
    );
  }
}

File: src/domain/usecases/SubmitReinforcementUseCase.ts
import { ILearnerRepository } from '../repositories/ILearnerRepository';
import { ILessonRepository } from '../repositories/ILessonRepository';
import { Question } from '../entities/Question';
import { SkillScoringEngine } from '../personalization/SkillScoringEngine';
import { ProgressEngine } from '../personalization/ProgressEngine';
import { RouteEngine } from '../personalization/RouteEngine';
import { LearningPath } from '../entities/LearningPathNode';

export interface ReinforcementResult {
  isCorrect: boolean;
  oldScore: number;
  newScore: number;
  xpAwarded: number;
  updatedRoadmap: LearningPath;
}

export class SubmitReinforcementUseCase {
  constructor(
    private learnerRepo: ILearnerRepository,
    private lessonRepo: ILessonRepository
  ) {}

  async execute(
    courseId: string,
    skillId: string,
    reinforcementNodeId: string,
    reinforcementQuestion: Question,
    selectedIndex: number
  ): Promise<ReinforcementResult> {
    const isCorrect = selectedIndex === reinforcementQuestion.correctIndex;
    const profile = await this.learnerRepo.getProfile();
    const courseScores = profile.scoresByCourse[courseId] || {};
    const currentSkillScore = courseScores[skillId]?.score ?? 0;

    // Record answer attempt in audit log
    await this.learnerRepo.recordAnswerAttempt({
      courseId,
      skillId,
      lessonId: reinforcementNodeId,
      questionId: reinforcementQuestion.id,
      selectedIndex,
      selectedAnswer: reinforcementQuestion.options[selectedIndex] || '',
      isCorrect,
    });

    if (!isCorrect) {
      const baseNodes = await this.lessonRepo.getBaseRoadmapNodes(courseId);
      const currentRoadmap = RouteEngine.adaptRoadmap(
        courseId,
        baseNodes,
        courseScores,
        profile.completedNodeIds,
        profile.completedReinforcementIds
      );
      return {
        isCorrect: false,
        oldScore: currentSkillScore,
        newScore: currentSkillScore,
        xpAwarded: 0,
        updatedRoadmap: currentRoadmap,
      };
    }

    // 1. Calculate new score deterministically
    const newScore = SkillScoringEngine.calculateReinforcementScore(currentSkillScore);

    // 2. Mark reinforcement node as completed
    await this.learnerRepo.markReinforcementCompleted(reinforcementNodeId);

    // Also mark the target node as completed so downstream unlocks
    const targetNodeId = reinforcementNodeId.replace('reinf_', '');
    await this.learnerRepo.markNodeCompleted(targetNodeId);

    // 3. Save new score
    const updatedSkillScore = {
      skillId,
      courseId,
      score: newScore,
      lastUpdated: Date.now(),
      masteryLevel: SkillScoringEngine.getMasteryLevel(newScore),
      isWeakestFocus: false,
    };

    await this.learnerRepo.saveSkillScores(courseId, {
      [skillId]: updatedSkillScore,
    });

    // 4. Calculate XP idempotently
    const { newXP, wasAwarded } = ProgressEngine.calculateNewXP(
      profile.xp,
      'reinforcement_completion',
      reinforcementNodeId,
      profile.completedReinforcementIds
    );

    if (wasAwarded) {
      await this.learnerRepo.updateProfile({ xp: newXP });
    }

    // 5. Adapt updated roadmap
    const updatedProfile = await this.learnerRepo.getProfile();
    const baseNodes = await this.lessonRepo.getBaseRoadmapNodes(courseId);
    const updatedScores = updatedProfile.scoresByCourse[courseId] || {};

    const updatedRoadmap = RouteEngine.adaptRoadmap(
      courseId,
      baseNodes,
      updatedScores,
      updatedProfile.completedNodeIds,
      updatedProfile.completedReinforcementIds
    );

    return {
      isCorrect: true,
      oldScore: currentSkillScore,
      newScore,
      xpAwarded: wasAwarded ? 30 : 0,
      updatedRoadmap,
    };
  }
}
